src/imitation/algorithms/preference_comparisons.py

```diff
@@ -0,0 +1,289 @@
+        batch_size: int = 32,
+    ):
+        super().__init__(model)
+        self.noise_prob = noise_prob
+        self.batch_size = batch_size
+        self.optim = th.optim.Adam(self.model.parameters())
+
+    def _loss(
+        self,
+        fragments: List[Tuple[TrajectoryWithRew, TrajectoryWithRew]],
+        preferences: np.ndarray,
+    ):
+        probs = th.empty(len(fragments), dtype=th.float32)
+        for i, fragment in enumerate(fragments):
+            frag1, frag2 = fragment
+            trans1 = flatten_trajectories_with_rew([frag1])
+            trans2 = flatten_trajectories_with_rew([frag2])
+            rews1 = self._rewards(trans1)
+            rews2 = self._rewards(trans2)
+            probs[i] = self._probability(rews1, rews2)
+        return th.nn.functional.binary_cross_entropy(
+            probs, th.as_tensor(preferences, dtype=th.float32)
+        )
+
+    def _rewards(self, transitions: TransitionsWithRew) -> th.Tensor:
+        return self.model(*self.model.preprocess(transitions))
+
+    def _probability(self, rews1: th.Tensor, rews2: th.Tensor) -> th.Tensor:
+        model_probability = 1 / (1 + (rews2 - rews1).sum().exp())
+        return self.noise_prob * 0.5 + (1 - self.noise_prob) * model_probability
+
+    def train(self, dataset: PreferenceDataset):
+        # TODO(ejnnr): This isn't specific to the loss function or probability model.
+        # In general, it mighe be best to split the probability model, the loss and
+        # the optimization procedure a bit more cleanly so that different versions
+        # can be combined
+
+        dataloader = th.utils.data.DataLoader(
+            dataset,
+            batch_size=self.batch_size,
+            sampler=th.utils.data.RandomSampler(
+                dataset, replacement=True, num_samples=None
+            ),
+            collate_fn=preference_collate_fn,
+        )
+        self.optim.zero_grad()
+        for fragments, preferences in dataloader:
+            loss = self._loss(fragments, preferences)
+            loss.backward()
+        self.optim.step()
+
+
+class PreferenceComparisons:
+    """Main interface for reward learning using preference comparisons."""
+
+    def __init__(
+        self,
+        agent_trainer: AgentTrainer,
+        reward_model: RewardNet,
+        agent_timesteps: int,
+        fragmenter: Optional[Fragmenter] = None,
+        preference_gatherer: Optional[PreferenceGatherer] = None,
+        reward_trainer: Optional[RewardTrainer] = None,
+    ):
+        """Initialize the preference comparison trainer.
+
+        Args:
+            agent_trainer: generates trajectories while optionally training an RL agent
+                on the learned reward function (can also be a sampler from a static
+                dataset of trajectories though).
+            reward_model: a RewardNet instance to be used for learning the reward
+            agent_timesteps: number of environment steps to train the agent for between
```

Comment:
Is this the number of timesteps it'll return as a sample? Do they always have to be the same?